---
title: "MD tag vs. CIGAR string in SAM/BAM file"
categories:
  - Blog
tags:
  - genomics
  - bioinformatics
  - notes
---

For my current project I had to parse information contained in the MD tag and CIGAR string which accompany each mapped read in a SAM/BAM file. I was confused intintially about the difference between MD and CIGAR, so I thought it might be helpful to summarise what I've learnt about MD and CIGAR string here.  

#### Tags
A SAM file stores aligned reads in a table format. Tags are optional fields containing [additional information about a read](https://www.samformat.info/sam-format-alignment-tags). A tag consists of three elements separated by `:`

1. Two letter TAG key, eg. MD, NM...  
2. A character indicating type of value stored. Available types:  

| Type  | Explained |
| ------------- | ------------- |
| A	| character |
| i	| signed 32-bit integer |
| f	| single-precision float |
| Z	| string |
| H | hex string |


3. TAG value  

For example, a tag in a SAM/BAM file `MD:Z:155` indicates that this is a `MD` tag containing string (`Z`) and taking value of `155`. 
See [SAM/BAM format specification](https://genome.sph.umich.edu/wiki/SAM 
https://samtools.github.io/hts-specs/SAMtags.pdf) for more information about different tags.  

#### MD tag
MD tag is a string encoding mismatched and deleted reference bases. It covers only aligned bases, which means it omits any insertions, soft clipped bases, padding, reference skips. MD tag is a progression of letters and characters representing *blocks* in the aligned sequence of a read. There are three types of *blocks* in the MD tag:

1. Integers, `[0 - 9]+` = a run of bases matching the reference sequence.   
2. Letters, usually `ACGT` = a single base that is different to the reference sequence, the letter indicates the nucleotide that doesnâ€™t match the reference. 
3. `^[Letters]` = a dash before the letters indicates a deletion in the aligned sequence. 

{% capture notice-2 %}
#### NOTE

* The non reference bases refer to a forward strand only as all aligned sequences in a BAM/SAM file are represented in forward direction. To find out whether a read mapped to a forward or reverse strand, look at the [FLAG](https://broadinstitute.github.io/picard/explain-flags.html) (the second column) 
{% endcapture %}

<div class="notice">{{ notice-2 | markdownify }}</div>

#### MD tag vs. CIGAR  
MD tag is not the same as CIGAR string. Although both details the concordance between a read and the reference sequence, CIGAR string takes the perspective of the read while MD tag is about the part aligned to the reference only.
This means that CIGAR contains the information about insertions and soft clipped bases. CIGAR indicates which part of a read was succesfully aligned, BUT provides no clue regarding whether this alignment has occured with or without any mismatches or deletions. Luckily, this information is stored in an MD tag. 

<img width="928" alt="Screenshot 2022-06-18 at 20 10 18" src="https://user-images.githubusercontent.com/32344189/174453024-b8a757fc-71af-47d0-8de2-67a7edabd09d.png">

#### Real-life example

Let's look at this read: 
```
K00178:339:HLV3FBBXY:4:1103:15919:16032	99	chr1	56375	0	73M	=	56375	73	GACTTATGTACTGCTCATTGTGTTGCAGAAGAGACCCCTGTGGGGGCAATAGAACAGATTTTCCTCTCACGTC	?????????????????????????????????????????????????????????????????????????	XA:Z:chr15,-101934533,73M,1;chr19,+97950,73M,1;	MC:Z:78S73M	MD:Z:36T36	PG:Z:MarkDuplicates	RG:Z:Sample1	NM:i:1	MQ:i:0	RQ:Z:FJA-FJF-A	UQ:i:32	AS:i:68	XS:i:68	RX:Z:CTTATGGTC
```
This is a mapped read with flag 99, which means that the read is paired (0x1), mapped in proper pair, first in pair, with mate on the reverse strand. 
```
read seq length: 73
CIGAR: 73M
MD: 36T36
```
It's fully aligned to the reference (CIGAR: `73M`) with one mismatch at position 37 (MD:`T`). A mismatch may indicate sequencing error or a true SNP/SNV at this genomic location. 

See [this post](https://vincebuffalo.com/notes/2014/01/17/md-tags-in-bam-files.html) for more examples.
